generators/app/templates/src/app/login/__auth._login.component.ts

```diff
@@ -40,16 +40,17 @@ export class LoginComponent implements OnInit {
 
   login() {
 <% if (props.ui === 'ionic') { -%>
-    const loading = this.loadingController.create();
-    loading.present();
+    const loadingPromise = this.loadingController.create();
+    const loadingPresentedPromise = loadingPromise
+      .then(loading => loading.present());
 <% } else { -%>
     this.isLoading = true;
 <% } -%>
     this.authenticationService.login(this.loginForm.value)
       .pipe(finalize(() => {
         this.loginForm.markAsPristine();
 <% if (props.ui === 'ionic') { -%>
-        loading.dismiss();
+        loadingPresentedPromise.then(() => loadingPromise.then(loading => loading.dismiss()));
```

Comment:
I think you could chain all the async work. You can transform a promise into an observable to keep the flow and also avoid saving intermediate promises.

could this work?:
```
const login$ = this.authenticationService.login(this.loginForm.value));
<% if (props.ui === 'ionic') { -%>
from(this.loadingController.create());
<% } else { -%>
login$
<% } -%>
    .pipe(
<% if (props.ui === 'ionic') { -%>
        map(loading => { this.loading = loading; loading.present() }),
        switchMap(() => login$
<% } -%>
        finalize(() => {
            this.loginForm.markAsPristine();
<% if (props.ui === 'ionic') { -%>
            this.loading.dismiss();
<% } -%>
        })
    );

Reply:
I tried it, and it worked great and looked great ( https://github.com/ngx-rocket/generator-ngx-rocket/blob/09e208569fc7fe47b22c73164e1afa053dc2ec47/generators/app/templates/src/app/login/__auth._login.component.ts )

...And all tests passed locally...

...And then e2e on CI started failing intermittently, and I couldn't reproduce it locally.

It looks like a bug in ionic to me that my quick and dirty promise code was somehow hiding, because there's no reason this version should be failing, because either way protractor is supposed to wait until the zone stabilizes, and both cases should stabilize only after BOTH the present() is done and the new component has fully loaded.

I couldn't investigate it further because I couldn't reproduce it, but something's gotta be up with the zone handling in either the loading controller or what they're doing with the router.

I'll try to find time to reproduce it with a simpler test case so I can submit it.

Anyway, I went back to the ugly version to confirm that this was indeed the variable.  I'll roll it back to the failing version later.

Reply:
FYI in case you friends are interested ;)

I was finally able to reproduce it locally with only promises and made a simple test case off of ionic's own official starters...  I'll be submitting this to ionic on Monday.

https://github.com/captaincaius/loadingtestcase/pull/1/files

Reply:
heh... looks like I pasted the PR diff, not the actual issue :P.

https://github.com/ionic-team/ionic/issues/15629

still no response yet on the issue.  I think I'll try to harass someone on ionic slack if I can't figure it out.  Debugging into minified transpiled ionic/core code is a 0/10 on the fun scale :P.

Reply:
comment added with link to commit to revert when loading controller works better
done :)